### PR TITLE
[pLz5YbCW] add checks for node labels and relationship types for apoc.merge procedures

### PR DESCRIPTION
**Pull request by nadja-muller**
apoc.merge.nodes:
- if labelNames null is newly considered as an empty label
- if labelNames contains null values, it is considered as incorrect input
- if labelNames contains among others an empty string, it is considered as incorrect input

apoc.merge.relationships:
- if relationship type is null, it is considered as incorrect input
- if relationship type is empty, it is considered as incorrect input

### CODE REVIEW

File: core/src/test/java/apoc/merge/MergeTest.java [hunk 1]
```diff
@@ -136,23 +124,59 @@ public void testMergeWithEmptyIdentityPropertiesShouldFail() {
             }
         }
     }
-    
+
+    @Test
+    public void testMergeNodeWithNullLabelsShouldFail() {
+        try {
+            testCall(db, "CALL apoc.merge.node([null, null], {name:'John'}) YIELD node RETURN node",
```
Comment (Lojjs) [inline core/src/test/java/apoc/merge/MergeTest.java:131]: It would be good to also have a test where not all labels are null, like `CALL apoc.merge.node(['Person', null], {name:'John'})...`

File: core/src/main/java/apoc/merge/Merge.java [hunk 2]
```diff
@@ -83,16 +82,20 @@ public Stream<NodeResultWithStats> nodeWithStats(@Name("labels") List<String> la
...
     private Result getNodeResult(List<String> labelNames, Map<String, Object> identProps, Map<String, Object> props, Map<String, Object> onMatchProps) {
-        if (identProps ==null || identProps.isEmpty()) {
+        if (identProps == null || identProps.isEmpty()) {
             throw new IllegalArgumentException("you need to supply at least one identifying property for a merge");
         }
 
-        String labels = labelString(labelNames);
+        if (labelNames != null && (labelNames.contains(null) || labelNames.contains(""))) {
+            throw new IllegalArgumentException("The list of label names may not contain any null or empty String values. If you wish to merge a node without a label, pass an empty list instead.");
+        }
```
Comment (Lojjs) [inline core/src/main/java/apoc/merge/Merge.java:91]: I don't know if we want to use the new types here like NODE, LIST and STRING, maybe check with @gem-neo4j as she has worked hard to unify the use of them.

Same applies to the relationship error message

File: core/src/test/java/apoc/merge/MergeTest.java [hunk 3]
```diff
@@ -136,23 +124,59 @@ public void testMergeWithEmptyIdentityPropertiesShouldFail() {
...
+        } catch (QueryExecutionException e) {
+            assertEquals(e.getMessage(), "Failed to invoke procedure `apoc.merge.node`: Caused by: java.lang.IllegalArgumentException: " +
+                    "The list of label names may not contain any null or empty String values. If you wish to merge a node without a label, pass an empty list instead.");
+        }
+    }
+
+    @Test
+    public void testMergeNodeContainingEmptyLabelShouldFail() {
+        try {
+            testCall(db, "CALL apoc.merge.node([''], {name:'John'}) YIELD node RETURN node",
```
Comment (Lojjs) [inline core/src/test/java/apoc/merge/MergeTest.java:155]: This test looks identical to the one above, I guess that you intended to send in '' rather than [''] here?

File: core/src/test/java/apoc/merge/MergeTest.java [hunk 4]
```diff
@@ -136,23 +124,59 @@ public void testMergeWithEmptyIdentityPropertiesShouldFail() {
...
+                    row -> assertTrue(row.get("node") instanceof Node));
+            fail();
+        } catch (QueryExecutionException e) {
+            assertEquals(e.getMessage(), "Failed to invoke procedure `apoc.merge.node`: Caused by: java.lang.IllegalArgumentException: " +
+                    "The list of label names may not contain any null or empty String values. If you wish to merge a node without a label, pass an empty list instead.");
+        }
+    }
+
+    @Test
+    public void testMergeNodeWithEmptyLabelListShouldFail() {
```
Comment (Lojjs) [inline core/src/test/java/apoc/merge/MergeTest.java:141]: It would be good to also have a test where not all labels are empty, like CALL apoc.merge.node(['Person' , ''], {name:'John'})...

**Review by Lojjs — CHANGES_REQUESTED** [review]
The check looks good but I have some comments on the tests and messages.

File: core/src/main/java/apoc/merge/Merge.java (hunk 2 above)
Comment (gem-neo4j) [reply core/src/main/java/apoc/merge/Merge.java]: I see the update and will add my comment here to continue the thread:

"The list of label names may not contain any `NULL` or empty `STRING` values. If you wish to merge a `NODE` without a label, pass an empty `LIST` instead."

The first list has been missed, and the second time list is mentioned, it mentions `LIST` by itself, which is not a valid Cypher Type name.

I am also unsure about our take on error messages containing personal pronouns?

Comment (Lojjs) [reply core/src/main/java/apoc/merge/Merge.java]: There are some new guidelines for error messages but we will discuss those in a meeting in 35 minutes so they are not really official/well structured yet

Comment (Lojjs) [reply core/src/main/java/apoc/merge/Merge.java]: This is WIP PR of error guidelines if you are curious 
https://github.com/neo4j/docs-status-codes/pull/73

Comment (nadja-muller) [reply core/src/main/java/apoc/merge/Merge.java]: How about this?
"The list of label names may not contain any NULL or empty STRING values. If you wish to merge a NODE without a label, pass an empty list instead."

Comment (Lojjs) [reply core/src/main/java/apoc/merge/Merge.java]: I think that is OK, do you agree @gem-neo4j ?

Comment (gem-neo4j) [reply core/src/main/java/apoc/merge/Merge.java]: Yeah, I think that is fine :)

**Review by Lojjs — APPROVED** [review]
I think it is fine now

**Review by Lojjs — APPROVED** [review]
Looks fine with the renaming too
